server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java

```diff
@@ -234,21 +240,74 @@ static SecureSettings loadSecureSettings(Environment initialEnv) throws Bootstra
             throw new BootstrapException(e);
         }
 
+        char[] passChars;
         try {
+            if (keystore != null && keystore.hasPassword()) {
+                passChars = readPassphrase(System.in, KeyStoreAwareCommand.MAX_PASSPHRASE_LENGTH);
+            } else {
+                passChars = new char[0];
+            }
+        } catch (IOException e) {
+            throw new BootstrapException(e);
+        }
+
+        try (SecureString password = new SecureString(passChars)) {
             if (keystore == null) {
                 final KeyStoreWrapper keyStoreWrapper = KeyStoreWrapper.create();
                 keyStoreWrapper.save(initialEnv.configFile(), new char[0]);
                 return keyStoreWrapper;
             } else {
-                keystore.decrypt(new char[0] /* TODO: read password from stdin */);
-                KeyStoreWrapper.upgrade(keystore, initialEnv.configFile(), new char[0]);
+                keystore.decrypt(password.getChars());
+                KeyStoreWrapper.upgrade(keystore, initialEnv.configFile(), password.getChars());
             }
         } catch (Exception e) {
             throw new BootstrapException(e);
         }
         return keystore;
     }
 
+    // visible for tests
+    /**
+     * Read from an InputStream up to the first carriage return or newline,
+     * returning no more than maxLength characters. This method discards characters
+     * beyond the newline, making it unsuitable for repeated readings from the stream.
+     */
+    static char[] readPassphrase(InputStream stream, int maxLength) throws IOException {
+        char[] buf = new char[maxLength + 2]; // potentially two terminating characters
+
+        try(InputStreamReader reader = new InputStreamReader(stream, StandardCharsets.UTF_8)) {
+            int len = 0;
+            int next;
+            while ((next = reader.read()) != -1) {
+                char nextChar = (char) next;
+                if (nextChar == '\n') {
```

Comment:
isn't this duplicating the read lines method in Terminal?

Reply:
Ah, right. Those two methods started out different but ended up on the same algorithm, with slightly different arguments and different exception messages. The method in bootstrap can delegate to the method in Terminal and preserve the different error messages.